### ISSUE TITLE
Role value does not persist after user tabs off combobox - MS Edge only?

### ISSUE THREAD

**Issue by mboeringa** [openstreetmap/iD#5449]
I only recently started adding turn restrictions using iD, but have noticed a disturbing issue. In the current setup, when you add a turn restriction, the textline for setting the role is editable: it allows typing. It even seems to support auto-completion.

As an advanced user will start to type the role, instead of waiting for any context menu to popup, iD suggests to support this, see the image:

![afbeelding](https://user-images.githubusercontent.com/7635726/47746029-6a7ae480-dc85-11e8-8699-716dd906b666.png)

However, confusingly, once you continue adding members (or at least think so!), and finally click the turn relation's link, you discover that while the members have been added, none of the roles have actually been set! If a user uploads the data like this, which is perfectly possible as iD doesn't block this, you end up with an incomplete and broken turn restriction relation in the database.

![afbeelding](https://user-images.githubusercontent.com/7635726/47746615-b417ff00-dc86-11e8-91ce-fb79e16f1ef2.png)

It was only after several tries, that I actually realized iD doesn't support setting roles via typing, but that you must wait for the context / role drop down menu with the preset roles to popup, before you can truly set the roles. **_Only roles chosen from the turn relation drop down preset, will actually be set on a turn restriction relation._**

![afbeelding](https://user-images.githubusercontent.com/7635726/47746473-71eebd80-dc86-11e8-83aa-cccf0998a780.png)

This is really confusing. I really suggest to dissalow editing and typing in the textbox, and possibly also to disallow uploading incomplete turn restriction relations to avoid polluting the OSM database with broken turn relations.

**Comment by bhousel** [comment]
I will check this - it sounds like the autocomplete is just running slow.

However I do encourage you to add turn restrictions in iD using the dedicated turn restriction editor, and not by manually editing the relations.

**Comment by mboeringa** [comment]
After looking through some more issues, I realize this is related or duplicate of https://github.com/openstreetmap/iD/issues/4900, that seems to describe the same issue for route relations (it may be a general issue with any relation type having preset roles).

**Comment by mboeringa** [comment]
> However I do encourage you to add turn restrictions in iD using the dedicated turn restriction editor, and not by manually editing the relations.

I actually start any new turn relation by clicking the + button for adding new relations, and then choosing Turn Restriction from the available options.

**Comment by bhousel** [comment]
> I actually start any new turn relation by clicking the + button for adding new relations, and then choosing Turn Restriction from the available options.

A much easier way to work with turn restrictions is to select any road junction and use the turn restriction editor in the sidebar.  You can select any incoming road and inspect and change the restrictions for any path through the intersection.  It has some nice built in help too.

![turn restrictions](https://user-images.githubusercontent.com/38784/47748372-00504880-dc61-11e8-912f-96ef10a907c6.gif)

**Comment by mboeringa** [comment]
@bhousel , this is important: I consistently see this fail in Microsoft Edge. I have now tried this in Firefox 63.0, and **it doesn't have the same issue**.

**Comment by mboeringa** [comment]
> A much easier way to work with turn restrictions is to select any road junction and use the turn restriction editor in the sidebar.

Thanks for the suggestion, but I don't see this specific relation editor popup when I click on the connection point between two roads?...

**Comment by mboeringa** [comment]
@bhousel 

Just curious, but were you able to reproduce this issue in Microsoft Edge? 

And note that the probably related issue #4900 also shows Safari 11.1 on macOS 10.12.6 being affected. The OP who reported that issue also confirmed my observation that Firefox is OK, and doesn't show the same issue (_"but not with Firefox 59.0.2 on macOS"_): https://github.com/openstreetmap/iD/issues/4900#issuecomment-377920274

**Comment by bhousel** [comment]
> Just curious, but were you able to reproduce the originally reported issue in Microsoft Edge?

No, sorry I haven't looked into this yet.

**Comment by bhousel** [comment]
I made some changes that I think should fix this.. Could you test in http://preview.ideditor.com/master/ and let me know if changing the role sticks now?

**Comment by bhousel** [comment]
I figured out what was going on with this one.  Fix coming soon..

**Comment by bhousel** [comment]
@mboeringa can you test this again in http://preview.ideditor.com/master/ ?
I think it's really fixed this time.

The issue was that some browsers don't always send a `change` event for changes that the user didn't do themselves (in this case, the autocomplete was doing the change).  We have workarounds all around the rest of the code to listen for both `change` and `blur` and see what actually changed - I just needed to add similar code to the role field too.

**Comment by mboeringa** [comment]
@bhousel 

I think you nailed it! The roles seem to stick now when typed in Edge. Going back and forth between relation members and the relation itself, the roles seem consistently kept.

It would be good to double check in Safari though, with which I can't help you.

![afbeelding](https://user-images.githubusercontent.com/7635726/49902142-67b60680-fe63-11e8-92a1-2bc79922ee7d.png)
